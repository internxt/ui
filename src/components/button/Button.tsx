import { ReactNode } from 'react';
import Spinner from '../spinner/Spinner';

interface ButtonProps {
  id?: string;
  variant?: 'primary' | 'secondary' | 'ghost' | 'destructive';
  type?: 'button' | 'submit';
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  onClick?: (e: unknown) => void;
  onKeyDown?: (e: unknown) => void;
  size?: 'medium' | 'default';
  loading?: boolean;
  dataTest?: string;
  autofocus?: boolean;
  buttonDataCy?: string;
  buttonChildrenDataCy?: string;
}

export const Button = ({
  variant = 'primary',
  type = 'button',
  id,
  children,
  className = '',
  disabled = false,
  onClick = () => undefined,
  onKeyDown = () => undefined,
  size = 'default',
  loading,
  dataTest,
  autofocus,
  buttonDataCy,
  buttonChildrenDataCy,
}: Readonly<ButtonProps>): JSX.Element => {
  let styles = '';

  if (variant === 'primary' && !disabled) {
    styles = `${loading ? 'bg-primary-dark' : 'bg-primary'} active:bg-primary-dark text-white shadow-sm`;
  } else if (variant === 'primary' && disabled) {
    styles = 'bg-gray-30 text-white shadow-sm';
  } else if (variant === 'destructive' && !disabled) {
    styles = `${loading ? 'bg-red-dark' : 'bg-red'} active:bg-red-dark text-white shadow-sm`;
  } else if (variant === 'destructive' && disabled) {
    styles = 'bg-gray-30 text-white shadow-sm';
  } else if (variant === 'secondary' && !disabled) {
    styles =
      'bg-surface dark:bg-gray-5 border border-gray-10' +
      ' hover:border-gray-20 active:bg-gray-1 dark:active:bg-gray-10 text-gray-80 shadow-sm';
  } else if (variant === 'secondary' && disabled) {
    styles = 'bg-surface dark:bg-gray-5 text-gray-30 border border-gray-5 shadow-sm';
  } else if (variant === 'ghost' && !disabled) {
    styles = 'hover:bg-gray-5 active:bg-gray-10 focus-visible:bg-gray-10';
  } else if (variant === 'ghost' && disabled) {
    styles = 'text-gray-30';
  }

  return (
    <button
      data-cy={buttonDataCy}
      id={id}
      onClick={onClick}
      onKeyDown={onKeyDown}
      disabled={disabled || loading}
      type={type}
      data-test={dataTest}
      autoFocus={autofocus}
      className={`${
        size === 'default' ? 'h-10 px-5' : 'h-8 px-3.5'
      } relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 ${styles} ${className}`}
    >
      {loading && <Spinner size={18} />}
      <div className="flex items-center justify-center space-x-2" data-cy={buttonChildrenDataCy}>
        {children}
      </div>
    </button>
  );
};
