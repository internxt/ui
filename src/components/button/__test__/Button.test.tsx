import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import Button from '../Button';
import renderer from 'react-test-renderer';

describe('Button component', () => {
  it('Button onClick should be called correctly', () => {
    const buttonClick = vi.fn();
    render(<Button variant="primary" onClick={buttonClick} />);
    const button = screen.getByRole('button');
    button.click();
    expect(buttonClick).toHaveBeenCalledOnce();
  });

  it('Primary button should render correctly', () => {
    const button = renderer.create(<Button variant="primary">Primary</Button>).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Secondary button should render correctly', () => {
    const button = renderer.create(<Button variant="secondary">Secondary</Button>).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Tertiary button should render correctly', () => {
    const button = renderer.create(<Button variant="tertiary">Tertiary</Button>).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Accent button should render correctly', () => {
    const button = renderer.create(<Button variant="accent">Accent</Button>).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Primary disabled button should render correctly', () => {
    const button = renderer.create(<Button variant="primary" disabled />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Secondary disabled button should render correctly', () => {
    const button = renderer.create(<Button variant="secondary" disabled />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Tertiary disabled button should render correctly', () => {
    const button = renderer.create(<Button variant="tertiary" disabled />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Accent disabled button should render correctly', () => {
    const button = renderer.create(<Button variant="accent" disabled />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Primary medium button should render correctly', () => {
    const button = renderer.create(<Button variant="primary" size="medium" />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Primary loading button should render correctly', () => {
    const button = renderer.create(<Button variant="primary" loading />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Primary submit button should render correctly', () => {
    const button = renderer.create(<Button variant="primary" type="submit" />).toJSON();
    expect(button).toMatchSnapshot();
  });
});
