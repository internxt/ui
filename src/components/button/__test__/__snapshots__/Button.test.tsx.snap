// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > Destructive button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-red active:bg-red-dark text-white shadow-sm "
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  >
    Destructive
  </div>
</button>
`;

exports[`Button component > Destructive disabled button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-gray-30 text-white shadow-sm "
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  />
</button>
`;

exports[`Button component > Primary button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-primary active:bg-primary-dark text-white shadow-sm "
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  >
    Primary
  </div>
</button>
`;

exports[`Button component > Primary disabled button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-gray-30 text-white shadow-sm "
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  />
</button>
`;

exports[`Button component > Primary loading button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-primary-dark active:bg-primary-dark text-white shadow-sm "
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <svg
    className="animate-spin "
    fill="none"
    height={18}
    viewBox="0 0 24 24"
    width={18}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      className="opacity-25"
      cx="12"
      cy="12"
      r="10"
      stroke="currentColor"
      strokeWidth="4"
    />
    <path
      className="opacity-75"
      d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824
              3 7.938l3-2.647z"
      fill="currentColor"
    />
  </svg>
  <div
    className="flex items-center justify-center space-x-2"
  />
</button>
`;

exports[`Button component > Primary medium button should render correctly 1`] = `
<button
  className="h-8 px-3.5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-primary active:bg-primary-dark text-white shadow-sm "
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  />
</button>
`;

exports[`Button component > Primary submit button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-primary active:bg-primary-dark text-white shadow-sm "
  onClick={[Function]}
  type="submit"
>
  <div
    className="flex items-center justify-center space-x-2"
  />
</button>
`;

exports[`Button component > Secondary button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-surface dark:bg-gray-5 border border-gray-10 hover:border-gray-20 active:bg-gray-1 dark:active:bg-gray-10 text-gray-80 shadow-sm "
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  >
    Secondary
  </div>
</button>
`;

exports[`Button component > Secondary disabled button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-surface dark:bg-gray-5 text-gray-30 border border-gray-5 shadow-sm "
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  />
</button>
`;

exports[`Button component > Ghost button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 hover:bg-gray-5 active:bg-gray-10 focus-visible:bg-gray-10 "
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  >
    Ghost
  </div>
</button>
`;

exports[`Button component > Ghost disabled button should render correctly 1`] = `
<button
  className="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 text-gray-30 "
  disabled={true}
  onClick={[Function]}
  type="button"
>
  <div
    className="flex items-center justify-center space-x-2"
  />
</button>
`;
